crates/gpui/src/elements/text.rs

```diff
@@ -305,11 +305,35 @@ impl TextLayout {
                 }
 
                 let mut line_wrapper = cx.text_system().line_wrapper(text_style.font(), font_size);
-                let text = if let Some(truncate_width) = truncate_width {
-                    line_wrapper.truncate_line(text.clone(), truncate_width, ellipsis)
+                let (text, truncated) = if let Some(truncate_width) = truncate_width {
+                    (
+                        line_wrapper.truncate_line(text.clone(), truncate_width, ellipsis),
+                        true,
+                    )
                 } else {
-                    text.clone()
+                    (text.clone(), false)
                 };
+                if truncated {
+                    let mut truncate_at = text.len() - 3;
+                    // let mut drop_unused_run = None;
+                    let mut run_index = 0;
+                    let mut iter = runs.iter_mut().peekable();
+                    while let Some(run) = iter.next() {
+                        if run.len <= truncate_at {
+                            truncate_at -= run.len;
+                        } else {
+                            run.len = truncate_at + 3;
+                            if iter.peek().is_some() {
+                                run_index += 1;
+                                break;
+                            }
+                        }
+                        run_index += 1;
+                    }
+                    if run_index < runs.len() {
+                        runs = runs[..run_index].to_vec();
+                    }
```

Comment:
This involves handling multiple `TextRun`s, and I'm not entirely sure if all cases are covered. The `3` here is hardcoded for now as a temporary solution. It's already quite late on my end, so I’ll address this tomorrow.

Reply:
Your right, here we need to update `runs` when `text` have truncated.

And found that the next `share_text` implementation, it looks like that handled the min line_text length to update run (I am not sure).

```rs
let run_len_within_line = cmp::min(line_end, run_start + run.len) - run_start;
```

https://github.com/zed-industries/zed/blob/b41ddbd0184d285400e1378172a9db9f480f95e5/crates/gpui/src/text_system.rs#L395

Reply:
I thought about it, there is looks like only need to update each `run.len` again. 

Take a look at my modification:

https://github.com/huacnlee/zed/pull/10/files

Reply:
> I thought about it, there is looks like only need to update each `run.len` again.
> 
> Take a look at my modification:
> 
> https://github.com/huacnlee/zed/pull/10/files

This is works, I have tested on Windows, the panic is gone, and the text renders is looks ok both on Windows and macOS.

Reply:
The idea here is that you may drop some runs after the truncation, I dont think just updating the `len` is enough, for the current simple case in `Extension` it works fine, not for other cases.